### PR TITLE
Add timestamp support to DBImplReadOnly

### PR DESCRIPTION
**Pull request by jowlyzhang**
This PR adds timestamp support to a read only DB instance opened as `DBImplReadOnly`. A follow up PR will add the same support to `CompactedDBImpl`.

 With this, read only database has these timestamp related APIs:

`ReadOptions.timestamp` : read should return the latest data visible to this specified timestamp
`Iterator::timestamp()` : returns the timestamp associated with the key, value
`DB:Get(..., std::string* timestamp)` : returns the timestamp associated with the key, value in `timestamp`

Test plan (on devserver):

```
$COMPILE_WITH_ASAN=1 make -j24 all
$./db_with_timestamp_basic_test --gtest_filter=DBBasicTestWithTimestamp.ReadOnlyDB*
```

### CODE REVIEW

File: db/db_impl/db_impl_readonly.h [hunk 1]
```diff
@@ -27,6 +27,9 @@ class DBImplReadOnly : public DBImpl {
   virtual Status Get(const ReadOptions& options,
                      ColumnFamilyHandle* column_family, const Slice& key,
                      PinnableSlice* value) override;
+  virtual Status Get(const ReadOptions& options,
```
Comment (riversand963) [inline db/db_impl/db_impl_readonly.h:30]: Nit: for new code, we can keep `override` and remove `virtual`, according to https://clang.llvm.org/extra/clang-tidy/checks/modernize-use-override.html.

File: db/db_impl/db_impl_readonly.cc [hunk 2]
```diff
@@ -33,20 +33,39 @@ DBImplReadOnly::~DBImplReadOnly() {}
...
+      return s;
+    }
+  } else {
+    const Status s = FailIfCfHasTs(column_family);
+    if (!s.ok()) {
+      return s;
+    }
   }
 
+  std::string* ts = column_family->GetComparator()->timestamp_size() > 0
```
Comment (riversand963) [inline db/db_impl/db_impl_readonly.cc:62]: Nit: to make the linter happy, I think we can do
```
const Comparator* const ucmp = column_family->GetComparator();
assert(ucmp);
std::string* ts = ucmp->timestamp_size() > 0 ? timestamp : nullptr;
```

**Review by riversand963 — COMMENTED** [review]
Thanks @jowlyzhang for the PR. Mostly LGTM with a few minor comments.

Can we add some test coverage for the error cases for `Get()`, `NewIterator()` and `NewIterators()`?
Furthermore, can you update HISTORY.md, or should we update it after updating `CompactedDBImpl` (in another PR)?

File: db/db_with_timestamp_basic_test.cc [hunk 3]
```diff
@@ -704,6 +704,158 @@ TEST_F(DBBasicTestWithTimestamp, SimpleIterate) {
   Close();
 }
 
+#ifndef ROCKSDB_LITE
```
[the remark this answers is not in the document]
Comment (jowlyzhang) [reply db/db_with_timestamp_basic_test.cc]: Good idea, the tests for read only cases are grouped into db_readonly_with_timestamp_test.cc now. I will add unit tests for CompactedDBImpl there too.

File: db/db_impl/db_impl_readonly.h (hunk 1 above)
Comment (jowlyzhang) [reply db/db_impl/db_impl_readonly.h]: Thanks for the pointer.

File: db/db_impl/db_impl_readonly.cc (hunk 2 above)
Comment (jowlyzhang) [reply db/db_impl/db_impl_readonly.cc]: Done

**Comment by jowlyzhang** [comment]
>

**Comment by jowlyzhang** [comment]
> unmerged



> Thanks @jowlyzhang for the PR. Mostly LGTM with a few minor comments.
> 
> Can we add some test coverage for the error cases for `Get()`, `NewIterator()` and `NewIterators()`? Furthermore, can you update HISTORY.md, or should we update it after updating `CompactedDBImpl` (in another PR)?

Thank you for the suggestion. I will update the HISTORY.md file in the follow up PR after `CompactedDBImpl` supports timestamp too.

**Review by riversand963 — COMMENTED** [review]
The test failure is legit and can be fixed by checking the return status.
Otherwise LGTM.

**Review by riversand963 — APPROVED** [review]
Thanks @jowlyzhang for the PR!

**Pull request closed without merging**